### PR TITLE
Bug 4370

### PR DESCRIPTION
**Pull request by LaurentBerger**
http://answers.opencv.org/question/62564/assertion-in-stitching_detailedcpp/

### CODE REVIEW

**Comment by boaz001** [comment]
Thanks for the PR!
One suggestion though: try not to change lines unrelated to the fix. This makes it harder to review because it not clear what change actually fixed it. I agree that spaces around '+' and after a ',' give better readability.
If you decide to combine both a fix and cosmetic code changes, at least create separate commits for both.

**Comment by LaurentBerger** [comment]
I'm sorry. I use VS 2013 editors and I fix option like this. Do you want me to  push source file without space around many separators?

**Comment by LaurentBerger** [comment]
I think there is another thing strange in this program it is about refine_mask( http://answers.opencv.org/question/64023/example-stitching_detailed-and-refine_mask/)
line 535 must be deleted

**Comment by alalek** [comment]
Please squash commits into one.

**Comment by alalek** [comment]
Please squash commits into one, because intermediate changes are not needed.
```
git rebase -i HEAD~2
```

**Comment by LaurentBerger** [comment]
It's impossible to make "git rebase -i HEAD~2 " something is lost on my new computer.
i proposed a new branch Bug4370a

**Comment by alalek** [comment]
after rebase you need to force git push (`git push -f origin Bug4370`)

**Comment by amroamroamro** [comment]
@LaurentBerger:
That doesn't look right... That part of the code is supposed to create and initialize the timelapser once if needed.

I think the test should be:

``` cpp
else if (!timelapser && timelapse)
{
    timelapser = Timelapser::createDefault(timelapse_type);
    timelapser->initialize(corners, sizes);
}
```

**Comment by LaurentBerger** [comment]
@amroamroamro Thanks for your comment. May be you're right. I have never used timelapser. I have detected this bug using this argument

**stitching_detailed "f:\images\pano\pano1.jpg" "f:\images\pano\pano2.jpg"**

with your modification and my argument it's OK I have got an image result.jpg

Now to test timelapser I think I have to use this argument :

**stitching_detailed "f:\images\pano\pano1.jpg" "f:\images\pano\pano2.jpg" --timelapse as_is 5**

Unfortunately I haven't got result but there is no exception. I think problem is at line https://github.com/Itseez/opencv/blob/master/samples/cpp/stitching_detailed.cpp#L828 string fixed cannot be append at begining of path. I have to rewrite some lines to have a good path.

As I don't understand last parameter timelapse_range i would appreciate your advise

**Comment by amroamroamro** [comment]
@LaurentBerger:

As I understand it, timelapsing will produce a sequence of "fixed" images.

The code is basic in building the new filenames, by simply prepending the word `fixed_` to existing filenames. As you found out, this will only work correctly if you have the images in the same working directory (`"fixed_C:\img.jpg"` is obviously nonsense!)

Ideally you would need to split filenames into three parts: base path, file name, and extension, and then you could add the prefix to the right part. I don't think there is a ready function in the C++ standard library to do that...

As an easy workaround, forget about the original names, and just write out numbered images like `fixed_1.jpg`, `fixed_2.jpg`, etc... using the existing loop counter `img_idx`.

**Comment by LaurentBerger** [comment]
I think it was that but it does not work. I have changed line  https://github.com/Itseez/opencv/blob/master/samples/cpp/stitching_detailed.cpp#L828 in : 
**imwrite("fixed_" + to_string(img_idx)+".jpg", timelapser->getDst());**

I have got only two images fixed_1.jpg and fixe_2.jpg. this two line are in loop  https://github.com/Itseez/opencv/blob/master/samples/cpp/stitching_detailed.cpp#L722.

value 5 is in varaible timelapse_range It is used only at https://github.com/Itseez/opencv/blob/master/samples/cpp/stitching_detailed.cpp#L469. I understand that's timelapse_range is bouding limit : 
example 100 images to stitch  image i is matched only with image i-5 to i+5 

About process namefile I will do it that's no a problem

@alalek I think there is something that  I do not understand in code. Merging is not ready (thanks @amroamroamro  for its comments)

**Comment by LaurentBerger** [comment]
@pglotov I have got some difficulties to understand Timelapse parameter in example Stitching_detailed.cpp. Why range_width parameter is used in BestOf2NearestRangeMatcher and not in class Timelapser?
thanks for your answers

**Comment by pglotov** [comment]
@LaurentBerger Timelapser processes images based on transformations computed by matcher, after matching is done. Matcher uses range_width to limit number of images to match with.

**Comment by LaurentBerger** [comment]
@alalek I think that merging is ready

**Comment by pglotov** [comment]
What is this for? Renaming parameters, adding new command line parameters, changing file name? Cleanup in the house?

**Comment by LaurentBerger** [comment]
As my english is not very good I'm not sure to understand your message. Is  your message polite or not
?

**Comment by pglotov** [comment]
Trying to understand the purpose of the changes. Could you please explain.

**Comment by LaurentBerger** [comment]
At the beginning of this PR it was only a bug with  ! (C++ not) at line 817. I think that it was only proble in this sample. @amroamroamro wrote to me about another problem (All it is written above). I have asked you to explain me how to use this parameter timelapse_range.

Now as you didn't answer to my question( polite or not) I think it was not polite. There is no problem i can delete this pull request you will make job yourself. I wait an answer from @alalek and he will decide

**Comment by pglotov** [comment]
stitching_detailed.cpp:816 is correct since if it is not a timelapse mode it creates a blender in the if clause on line 795, and otherwise it creates a timelapser. As to your question of politeness my question to you was polite and I wanted to understand why you decided to change formats of auxiliary parameters and change interface. Interface change should be only done for a good reason since it may cause a lot of additional work for someone already using current interface.

File: samples/cpp/stitching_detailed.cpp [hunk 1]
```diff
@@ -824,8 +828,13 @@ int main(int argc, char* argv[])
...
             timelapser->process(img_warped_s, Mat::ones(img_warped_s.size(), CV_8UC1), corners[img_idx]);
-
-            imwrite("fixed_" + img_names[img_idx], timelapser->getDst());
+            String fileName(img_names[img_idx]);
+            size_t posExt=fileName.find_last_of('.');
+            if (posExt!=string::npos)
+            {
+                fileName = fileName.substr(0,posExt)+"_fixed"+fileName.substr(posExt,fileName.length()-posExt);
+                imwrite(fileName, timelapser->getDst());
+            }
```
Comment (alalek) [inline samples/cpp/stitching_detailed.cpp]: Agreed for cases like "C:\path\image_000x.jpg", but it is better to preserve existed behavior.
Something like this (code is not checked actually):
```
size_t posFileName = std::max(fileName.find_last_of('/') + 1, fileName.find_last_of('\\') + 1); // don't optimize +1
fileName = fileName.substr(0, posFileName) + "fixed_" + fileName.substr(posFileName, fileName.length() - posFileName);
imwrite(fileName, timelapser->getDst());
```

File: samples/cpp/stitching_detailed.cpp [hunk 2]
```diff
@@ -813,7 +817,7 @@ int main(int argc, char* argv[])
             }
             blender->prepare(corners, sizes);
         }
-        else if (!timelapser)
+        else if (!timelapser && timelapse)
```
Comment (alalek) [inline samples/cpp/stitching_detailed.cpp]: Right. Agreed with fix.
Lines 795+816 works as expected, but it is for first iteration only (there is a for loop). Application with disabled timelapse mode was assert message.

File: samples/cpp/stitching_detailed.cpp [hunk 3]
```diff
@@ -120,7 +120,9 @@ static void printUsage()
         "  --output <result_img>\n"
         "      The default is 'result.jpg'.\n"
         "  --timelapse (as_is|crop) (range_width)\n"
-        "      Output warped images separately as frames of a time lapse movie, with 'fixed_' prepended to input file names.\n";
+        "      Output warped images separately as frames of a time lapse movie, with 'fixed_' appended before extension input file names.\n"
+        "  --rangewidth <int>\n"
+        "      uses range_width to limit number of images to match with.\n";
```
Comment (alalek) [inline samples/cpp/stitching_detailed.cpp]: This changes makes sense because OpenCV CommandLineParser can process only one parameter value at once (There is somewhere a PR with samples migration to CommandLineParser).

- `--timelapse (as_is|crop) (range_width)`: need to fix message and remove " (range_width)"
- my suggestion is to rename rangewidth to timelapse_range/timelapse_rangewidth (because it is used only with timelapse)
- `fixed_`/`_fixed` message (see the separate note below)

File: samples/cpp/stitching_detailed.cpp [hunk 4]
```diff
@@ -326,8 +328,10 @@ static int parseCmdArgs(int argc, char** argv)
                 return -1;
             }
             i++;
-
-            timelapse_range = atoi(argv[i + 1]);
+        }
+        else if (string(argv[i]) == "--rangewidth")
+        {
+            range_width = atoi(argv[i + 1]);
```
Comment (alalek) [inline samples/cpp/stitching_detailed.cpp]: My suggestion is to add `timelapse = true;` somewhere here in case of renamed parameter.

File: samples/cpp/stitching_detailed.cpp (hunk 2 above)
Comment (pglotov) [inline samples/cpp/stitching_detailed.cpp]: I see, this should work. However, can this if-else combo be taken out of the image loop? Looks like it is independent of images and is executed only once.

Comment (alalek) [inline samples/cpp/stitching_detailed.cpp]: This sample is rather old and big (~1k lines). A log of changes have been added here (includes possible caching for Blender object, which looks curiously now).
I believe it will be good to have separate PRs with new changes (this PR already has many fixes)

Comment (LaurentBerger) [inline samples/cpp/stitching_detailed.cpp]: If you mean that there is too many parameters in this sample I 'm 
agree.  I propose to close this pull request and make two new pull 
request one without timelapser and another with timelapser parameter 
only (with fixed parameter only). If you are agree I can manage first 
example and may be  @*pglotov <https://github.com/pglotov> * can manage 
other PR

Comment (alalek) [inline samples/cpp/stitching_detailed.cpp]: I mean that it not easy to understand old samples with 1k lines for new users like @pglotov, you or me. There are many historical/legacy problems.

**Comment by LaurentBerger** [comment]
@alalek @pglotov I have worked on this sample and I think really good way now is  to make two examples one for stitching and another for timelapser.

File: samples/cpp/stitching_detailed.cpp [hunk 5]
```diff
@@ -120,7 +120,9 @@ static void printUsage()
         "  --output <result_img>\n"
         "      The default is 'result.jpg'.\n"
         "  --timelapse (as_is|crop) (range_width)\n"
-        "      Output warped images separately as frames of a time lapse movie, with 'fixed_' prepended to input file names.\n";
+        "      Output warped images separately as frames of a time lapse movie, with 'fixed_' appended before extension input file names.\n"
```
Comment (sturkmen72) [inline samples/cpp/stitching_detailed.cpp]: i think no need to change this line

**Comment by alalek** [comment]
@LaurentBerger Well done, it works!

Could you please squash commits into one?
There are some merge commits, so conflicts are raised on classical "git rebase", so lets get help from GitHub:
```
# 1) update OpenCV master branch (upstream/master should be updated)
git fetch upstream master
# 2) backup current branch (optional)
git branch Bug4370_old Bug4370
# 3) get merged changes via GitHub
git fetch upstream +pull/4110/merge:pr4110_merge
# 4) reset changes
git checkout -B Bug4370 upstream/master
# 5) apply merged changes
git checkout pr4110_merge -- .
# 6) commit changes
git commit -m "fixed #5005: stitching_detailed.cpp"
# 7) update PR
git push -f origin Bug4370
```

File: samples/cpp/stitching_detailed.cpp [hunk 6]
```diff
@@ -813,9 +817,8 @@ int main(int argc, char* argv[])
             }
             blender->prepare(corners, sizes);
         }
-        else if (!timelapser)
+        else if (!timelapser && timelapse)
         {
-            CV_Assert(timelapse);
```
Comment (pglotov) [inline samples/cpp/stitching_detailed.cpp]: Why remove this line? It can get here only in timelapse mode.

File: samples/cpp/stitching_detailed.cpp [hunk 7]
```diff
@@ -119,8 +119,10 @@ static void printUsage()
         "      Blending strength from [0,100] range. The default is 5.\n"
         "  --output <result_img>\n"
         "      The default is 'result.jpg'.\n"
-        "  --timelapse (as_is|crop) (range_width)\n"
-        "      Output warped images separately as frames of a time lapse movie, with 'fixed_' prepended to input file names.\n";
+        "  --timelapse (as_is|crop) \n"
+        "      Output warped images separately as frames of a time lapse movie, with 'fixed_' prepended to input file names.\n"
+        "  --rangewidth <int>\n"
+        "      uses range_width to limit number of images to match with.\n";
 }
```
Comment (pglotov) [inline samples/cpp/stitching_detailed.cpp]: Now rangewidth has become independent parameter. I'd prefer to leave it as is: all timelapse related parameters are grouped together.

File: samples/cpp/stitching_detailed.cpp (hunk 6 above)
Comment (alalek) [inline samples/cpp/stitching_detailed.cpp]: "assert" after "if" with the same check is redundant:
```
if (!timelapser && timelapse)
```

File: samples/cpp/stitching_detailed.cpp (hunk 7 above)
Comment (alalek) [inline samples/cpp/stitching_detailed.cpp]: Please look at this comment about OpenCV CommandLineParser: 
https://github.com/Itseez/opencv/pull/4110#discussion_r42940259

**Comment by alalek** [comment]
:+1:

**Comment by LaurentBerger** [comment]
thanks you very much @alalek for github recipes
